Add Gazell support (#5281)
nrfconnect/sdk-nrf: doc/nrf/ug_gzp.rst

The Received Signal Strength Indication (RSSI) is used on the Host side of the nRF5 so that only Devices residing close (<30 cm) to the Host are allowed to pair (on nRF24L Host nodes, the carrier detect functionality is employed).
Proximity estimation is essential to avoid confusion when pairing in certain environments, for example in an office.

Pairing and key exchange stages
*******************************

The pairing and key exchange process consists of the following stages:

* :ref:`ug_gzp_sys_addr_exchange`
* :ref:`ug_gzp_host_id_exchange`
* :ref:`ug_gzp_dyn_key_exchange`
* :ref:`ug_gzp_enc_user_data_exchange`

Each of these stages are required to send encrypted user data from a Device to a Host.
If encryption is not required, only the "System Address exchange" stage is required to send unencrypted data on pipes 2-7.

Factory preset parameters
=========================

For pairing and key exchange to occur, an unpaired Host and Device must share the following factory preset parameters:

+----------------------+---------+-------+--------------------------------------------------+----------------+

> | Parameter            | Secret? | Bytes | Purpose                                          | Set            |

Reviewer: ```suggestion
| Parameter            | Secret  | Bytes | Purpose                                          | Set            |
```

+======================+=========+=======+==================================================+================+
| Secret Key           | Yes     | 16    | Encrypting Host ID when sent from Host to Device | Factory preset |
+----------------------+---------+-------+--------------------------------------------------+----------------+
| Pairing Address      | No      | 5     | Sending of System Address from Host to Device    | Factory preset |
+----------------------+---------+-------+--------------------------------------------------+----------------+
| Packet Validation ID | No      | 3     | Authentication of packets                        | Factory preset |
+----------------------+---------+-------+--------------------------------------------------+----------------+

In addition, every Host has the unique parameters that are generated at runtime:

+----------------+---------+-------+------------------------------------------------------+-------------------------------------------------------+

> | Parameter      | Secret? | Bytes | Purpose                                              | Set                                                   |

Reviewer: ```suggestion
| Parameter      | Secret  | Bytes | Purpose                                              | Set                                                   |
```

+================+=========+=======+======================================================+=======================================================+
| System Address | No      | 5     | Address used for all transmission.                   | Runtime.                                              |
|                |         |       | Seed for generating channel set.                     | Generated when first system address request received. |
+----------------+---------+-------+------------------------------------------------------+-------------------------------------------------------+
| Host ID        | Yes     | 5     | Encrypting Dynamic Key when sent from Device to Host | Runtime.                                              |
|                |         |       |                                                      | Generated when first Host ID request is received.     |
+----------------+---------+-------+------------------------------------------------------+-------------------------------------------------------+

> The System Address and Host ID are stored in non-volatile memory (NVM) and apply for the lifetime of the Host unless the NVM is erased.

Reviewer: ```suggestion
The System Address and Host ID are stored in the non-volatile memory (NVM) and apply for the lifetime of the Host unless the NVM is erased.
```

.. figure:: images/gzp_factory_defaults.svg
   :alt: Initial pairing parameters

   Initial pairing parameters

.. _ug_gzp_sys_addr_exchange:

System Address exchange
=======================

The System Address is an individual Host's private address.
For a Device to communicate with an individual Host, it needs to know the Host's System Address.
When requesting a System Address, the Device sets the TX power to a low level using GZP_POWER.
By measuring the received signal strength when a pairing request is received from a Device, the Host can estimate the relative proximity of the requesting Device.

For a Device to pair with a Host, it must first obtain the System Address on which all subsequent key exchange and data transfer occur.
This transaction occurs on the pipe 0 and is transmitted in cleartext on the air as it is not a secret.

.. figure:: images/gzp_address_exchange.svg
   :alt: System address exchange

   System Address exchange

.. _ug_gzp_host_id_exchange:

Host ID exchange
================

Once the Device has the System Address, it can request the Host ID on the pipe using GZP_DATA_PIPE.

> The Host ID is used to generate subsequent "dynamic keys" for encrypted data transfer.

Reviewer: ```suggestion
The Host ID is used to generate subsequent dynamic keys for encrypted data transfer.
```

After receiving a Host ID request, the Host generates the ID if it has not done it already.
The Host ID is generated using the random Session Token received from the Device in the Host ID request as well as the session counter.

The Device sends a packet to fetch the Host ID.
The secret Host ID is transmitted on the encrypted pipe GZP_DATA_PIPE using the shared Secret Key.

The following security precautions are taken for the Host ID exchange:

* Passive eavesdropping is prevented by using AES encryption.
* Replay attacks are prevented by using session tokens.

> * Man-in-the-middle and malicious device attacks can be prevented by implementing a user validation stage before the Host ID is sent to the Device (see :ref:`gazell_host_validation`).

Reviewer: ```suggestion
* Man-in-the-middle and malicious device attacks can be prevented by implementing a user validation stage before the Host ID is sent to the Device (see :ref:`ug_gzp_host_id_validation`).
```

.. figure:: images/gzp_host_id_exchange.svg
   :alt: Host ID exchange

   Host ID exchange

The Host ID can be compromised if the attacker has knowledge of the Secret Key.
The attacker could eavesdrop the Host ID exchange and obtain the Host ID or attempt to pair as a malicious device and obtain the Host ID.
Before the Host ID is sent from the Host to the Device, the application can add a validation stage.

> .. _gazell_host_validation

Reviewer: ```suggestion
.. _gazell_host_validation:
```
My bad, sorry.

Author: The label is renamed and a colon is added.

Reviewer: When you rename the label, you also need to change the link to this label. I commented this accordingly where the actual link is.

* Only the Device used for sending user data during the validation stage can decrypt the Host ID sent from the Host.

.. _ug_gzp_dyn_key_exchange:

Dynamic Key exchange
====================

The Dynamic Key is used for encrypting user data.
Each Device must have a unique Dynamic Key, and the Host must know the Dynamic Key of each Device it communicates with.

A Device can initialize the update of the Dynamic Key at any time.
The Dynamic Key is generated randomly on the Device and then communicated to the Host.
The Host ID is used for encrypting the Dynamic Key.

The Dynamic Key is considered a secret, and the following security precautions are taken:

* Passive eavesdropping is prevented by using AES encryption.
* Replay attacks are prevented by using session tokens sent from Host.
* Only Devices knowing the Host ID can update the Dynamic Key in the Host.

.. figure:: images/gzp_key_exchange.svg
   :alt: Dynamic Key exchange

   Dynamic Key exchange

The main reasons for using a Dynamic Key for encryption of user data are:

* A Host must be able to pair with multiple Devices at the same time, and none of these should use the same key for encryption of user data.

> * The solution must be available on OTP devices, where storing of keys in non-volatile memory during runtime is not desired.

Reviewer: ```suggestion
* The solution must be available on OTP devices, where storing of keys in the non-volatile memory during runtime is not desired.
```

The secrecy of the Dynamic Key is dependent on the secrecy of the Host ID.
The Dynamic Key can be compromised if both these conditions are met:

* Attacker eavesdrops the exchange of the Dynamic Key.
* The Host ID has been compromised.

.. _ug_gzp_enc_user_data_exchange:

Encrypted User Data exchange
****************************

Once the Device and Host share a Dynamic key, encrypted data exchange can occur.
When sending Encrypted User Data, the following security precautions are taken:

> * Passive eavesdropping prevented by AES encryption.
> * "Hostile" Device attacks prevented as only Devices knowing the current Dynamic Key can send user data that will be accepted by the Host.
> * Known plaintext or ciphertext attacks prevented by AES encryption.
> * Replay attacks prevented by using session tokens sent from the Host.
> * Frequency analysis attacks prevented by updating session token for every packet.

Reviewer: ```suggestion
* Passive eavesdropping is prevented by AES encryption.
* Hostile device attacks are prevented as only Devices that know the current Dynamic Key can send user data that will be accepted by the Host.
* Known plaintext or ciphertext attacks are prevented by AES encryption.
* Replay attacks are prevented by using session tokens sent from the Host.
* Frequency analysis attacks are prevented by updating session token for every packet.
```

.. figure:: images/gzp_user_data_exchange.svg
   :alt: Encrypted user data exchange

   Encrypted user data exchange

The secrecy of the Encrypted User Data is dependent on the secrecy of the Dynamic Key.
The Encrypted User Data may be compromised if both these conditions are met:

* Attacker eavesdrops the user data exchange.
* The current Dynamic Key has been compromised.

Resource requirements
*********************

In addition to the resources required by the :ref:`ug_gzll` (see :ref:`ug_gzll_resources`), Gazell Pairing also employs three nRF5 peripherals:

* Random Number Generator, for generating keys and tokens.
* AES Electronic Codebook (ECB), for encryption and decryption.
* Non-Volatile Memory Controller (NVMC), for storing of pairing parameters.

> In addition Gazell Pairing employs the following Gazell Link Layer resources:

Reviewer: ```suggestion
In addition, Gazell Pairing employs the following Gazell Link Layer resources:
```

* Two pipes: one for pairing and one for encrypted data transmission.
* Gazell pairing determines the channel set used by Gazell.

Since GZP requires exclusive access to pipes 0 and GZP_DATA_PIPE (default pipe 1), it must control the internal Gazell Link Layer variables ``base_address_0``, ``base_address_1`` and ``prefix_address_byte`` for pipes GZP_PAIRING_PIPE (always pipe 0) and GZP_DATA_PIPE (configurable).
The main application can use the pipes 2-7.
The ``base_address_1`` applies to these pipes.
Gazell Pairing must also determine whether the RX pipes 0 and 1 are enabled.
Make sure not to affect the ``rx_enabled`` status of these pipes.

The following Gazell Link Layer API functions should not be accessed:

* ``nrf_gzll_set_base_address_0()``
* ``nrf_gzll_set_base_address_1()``
* ``nrf_gzll_set_address_prefix_byte()`` (for pipes 0 and 1)
* ``nrf_gzll_set_rx_pipes_enabled()`` (can be used but the enabled status of pipes 0 and 1 should not be modified)
* ``nrf_gzll_set_channel_table()``

> System Address and Channel Table generation

Reviewer: ```suggestion
System Address and channel table generation
```
